keras/src/quantizers/gptqutils.py

```diff
@@ -0,0 +1,386 @@
+import io
+import logging
+import random
+import tarfile
+import time
+
+import numpy as np
+import requests
+from datasets import load_dataset
+
+from keras.src import ops
+from keras.src.layers import Dense
+from keras.src.layers import EinsumDense
+from keras.src.layers import Embedding
+
+from .gptq import GPTQ
+from .quant import Quantizer
+
+
+def get_dataloader(tokenizer, seqlen, dataset, nsamples=128, seed=0):
+    """
+    Prepares and chunks the calibration dataloader, repeating short datasets.
+    """
+    all_tokens = []
+
+    # --- Step 1: Unify all input types into a single list of tokens ---
+    if isinstance(dataset, str):
+        logging.info(f"Loading '{dataset}' dataset from Hub...")
+        if dataset == "wikitext2":
+            d_name, d_config = "wikitext", "wikitext-2-raw-v1"
+        elif dataset == "ptb":
+            url = "https://www.fit.vutbr.cz/~imikolov/rnnlm/simple-examples.tgz"
+            try:
+                # 1. Download the archive into memory
+                response = requests.get(url)
+                response.raise_for_status()
+
+                # 2. Extract only the test file from the in-memory archive
+                with tarfile.open(
+                    fileobj=io.BytesIO(response.content), mode="r:gz"
+                ) as tar:
+                    train_path = "./simple-examples/data/ptb.train.txt"
+                    test_bytes = tar.extractfile(train_path).read()
+
+                # 3. Decode the bytes and join into a single string
+                test_lines = test_bytes.decode("utf-8").strip().split("\n")
+                full_text = "\n\n".join(test_lines)
+                all_tokens = tokenizer.tokenize(full_text)
+                logging.info(
+                    "✅ Successfully processed PTB training data for"
+                    "calibration."
+                )
+
+                # 2. Perform sampling and chunking directly inside this block
+                all_tokens = np.array(all_tokens, dtype=np.int32)
+                required_tokens = nsamples * seqlen
+                if len(all_tokens) < required_tokens:
+                    logging.info(
+                        f"Warning: PTB dataset is too short ({len(all_tokens)}"
+                        "tokens). Repeating data."
+                    )
+                    repeats = -(-required_tokens // len(all_tokens))
+                    all_tokens = np.tile(all_tokens, repeats)
+
+                calibration_samples = []
+                for _ in range(nsamples):
+                    start_index = random.randint(
```

Comment:
use rng instead of random package: https://numpy.org/devdocs/reference/random/generator.html